clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -642,6 +658,39 @@ class FindPFWGLambdaFnVisitor
   const CXXRecordDecl *LambdaObjTy;
 };
 
+// Searches for a call to PF lambda function and captures it.
+class FindPFLambdaFnVisitor
+    : public RecursiveASTVisitor<FindPFLambdaFnVisitor> {
```

Comment:
Could you please provide a code example which we are searching for in this visitor? I'm not sure that I fully get what we are doing here.

Reply:
This looks like it is searching for a parallel-for call by checking number of arguments, type of arguments, etc. But I guess my understanding is incorrect because we don't need to traverse the kernel body for this? Or is this traversal required because of the transformation mentioned above?

Reply:
See my reply to Mariya's comment. If an attribute "this kernel calls this_item" could be defined and used for this purpose, much of the new SemaSYCL code would be unnecessary.
Would that be an acceptable alternate solution? 
The current solution does not keep this info around. It is computed/discovered and immediately used during integration header generation.

Reply:
I implemented a scheme where an internal attribute "callsThisItem" is introduced and leveraged the existing call-graph traversal to note calls to this_item and mark the kernel with that attribute. Unfortunately I discovered that the integration header is generated *before* this walk of the CG. So this scheme will not work.
We have to stay with this scan of the parallel_for kernel callers during integration header generation to detect use of this_item and mark the integration header entry accordingly.

Reply:
Then my question stays the same, could you please describe a code example which is searched by this visitor?

Reply:
The test parallel_for_this_item.cpp included in this PR is an example.
The scan looks for "this_item" called from the kernel.

Reply:
So, this visitor searches for kernel object call operator and there is nothing special?
In SemaSYCL we have a kernel object type as `CXXRecordDecl`. Can we just walk through its methods and save a call operator? For example using this https://github.com/intel/llvm/blob/4326b95635756885010632c033da2b546e8084a5/clang/lib/Sema/SemaSYCL.cpp#L2067 function.

Reply:
Correct me if I'm mistaken, but if he peeks through the CXXRecordDecl he is only looking at the kernel itself, no? Meaning, not any sub-functions it might call. 
The user may be calling `this_item()` from their kernel function directly, or possibly by one of the sub-functions it calls.

Reply:
The search starts at a parallel_for kernel and looks through the call-graph at all the callees.
So this will also be detected as a call to this_item:

SYCL_EXTERNAL cl::sycl::item<1> f() { return cl::sycl::this_item<1>(); }
...
cgh.parallel_for<class C>(cl::sycl::range<1>(1), [=](cl::sycl::id<1> I) { f(); });

Reply:
I've updated the test parallel_for_this_item.cpp to show use of this_item() in functions called from the kernel.

Reply:
I was speaking only about the `FindPFLambdaFnVisitor` Visitor, for me it looks unnecessary, since it looks like it searches only for kernel object call operator. My understanding was that recursive search through callgraph happens in `setThisItemIsCalled`, so If we remove `FindPFLambdaFnVisitor` and use simpler method of getting kernel object call operator, we don't break recursive search.

Reply:
Yes, you are right that getOperatorParens can be used here. I've made that change. That eliminates the need for "class FindPFLambdaFnVisitor".

A similar situation exists with "class FindPFWGLambdaFnVisitor" and its one use. However, that is existing code unrelated to this change so I'm leaving it as-is.

Reply:
Great, thanks! 

> A similar situation exists with "class FindPFWGLambdaFnVisitor" and its one use. However, that is existing code unrelated to this change so I'm leaving it as-is.

Yeah, I think it is not in scope of this patch.